File: medium/2300. Successful Pairs of Spells and Potions/2300. Successful Pairs of Spells and Potions.py
from typing import List


class Solution:
    def successfulPairs(
        self, spells: List[int], potions: List[int], success: int
    ) -> List[int]:
        potions.sort()
        res = []

        for s in spells:
            biggest = potions[-1] * s

            if biggest < success:
                res.append(0)
                continue

            left, right = 0, len(potions) - 1

            while left < right:
                mid = left + (right - left) // 2
                if potions[mid] * s > success:
                    right = mid
                else:
                    left = mid + 1

            res.append(len(potions) - left)

        return res


solution = Solution()
print(solution.successfulPairs(spells=[5, 1, 3], potions=[1, 2, 3, 4, 5], success=7))

File: medium/2300. Successful Pairs of Spells and Potions/2300. Successful Pairs of Spells and Potions.go
package main

import (
	"slices"
)

func successfulPairs(spells []int, potions []int, success int64) []int {
	n := len(potions)
	slices.Sort(potions)
	p := []int{}

	for _, s := range spells {
		biggest := potions[n-1] * s

		if biggest < int(success) {
			p = append(p, 0)
			continue
		}

		left, right := 0, n-1
		for left < right {
			mid := (right + left) / 2
			if potions[mid]*s < int(success) {
				left = mid + 1
			} else {
				right = mid
			}
		}

		p = append(p, n-left)
	}

	return p
}
